File: server/src/routes/registry.js
import express from 'express';
import { getDb } from '../config/db.js';

const router = express.Router();

// --- Vehicles ---
router.get('/vehicles', async (req, res) => {
  try {
    const db = getDb();
    const rows = await db.all('SELECT * FROM vehicles');
    res.json({ status: 'SUCCESS', data: rows });
  } catch (err) {
    res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

router.post('/vehicles', async (req, res) => {
  try {
    const { id = `v-${Date.now()}`, name, registration_number, device_id, vehicle_type, fuel_type, driver_name, blood_group, insurance_policy } = req.body;
    const db = getDb();
    await db.run(`
      INSERT INTO vehicles (id, name, registration_number, device_id, vehicle_type, fuel_type, driver_name, blood_group, insurance_policy)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `, [id, name, registration_number, device_id, vehicle_type, fuel_type, driver_name, blood_group, insurance_policy]);
    res.status(201).json({ status: 'SUCCESS', message: 'Vehicle registered' });
  } catch (err) {
    res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

// --- Medical Profile ---
router.get('/medical', async (req, res) => {
  try {
    const db = getDb();
    const profile = await db.get('SELECT * FROM medical_profiles LIMIT 1');
    res.json({ status: 'SUCCESS', data: profile });
  } catch (err) {
    res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

router.put('/medical', async (req, res) => {
  try {
    const { full_name, age, gender, blood_group, abha_id, emergency_notes, allergies, medical_conditions, primary_physician_name, primary_physician_phone, organ_donor } = req.body;
    const db = getDb();
    await db.run(`
      UPDATE medical_profiles 
      SET full_name = ?, age = ?, gender = ?, blood_group = ?, abha_id = ?, emergency_notes = ?, allergies = ?, medical_conditions = ?, primary_physician_name = ?, primary_physician_phone = ?, organ_donor = ?, updated_at = CURRENT_TIMESTAMP
      WHERE id = 'med-1'
    `, [full_name, age, gender, blood_group, abha_id, emergency_notes, allergies, medical_conditions, primary_physician_name, primary_physician_phone, organ_donor ? 1 : 0]);
    res.json({ status: 'SUCCESS', message: 'Medical profile updated' });
  } catch (err) {
    res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

// --- Emergency Contacts ---
router.get('/contacts', async (req, res) => {
  try {
    const db = getDb();
    const rows = await db.all('SELECT * FROM emergency_contacts ORDER BY is_primary DESC');
    res.json({ status: 'SUCCESS', data: rows });
  } catch (err) {
    res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

router.post('/contacts', async (req, res) => {
  try {
    const { id = `ec-${Date.now()}`, user_id = 'u1', name, relation, phone, is_primary = 0, notify_sms = 1 } = req.body;
    const db = getDb();
    await db.run(`
      INSERT INTO emergency_contacts (id, user_id, name, relation, phone, is_primary, notify_sms)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `, [id, user_id, name, relation, phone, is_primary ? 1 : 0, notify_sms ? 1 : 0]);
    res.status(201).json({ status: 'SUCCESS', message: 'Contact added' });
  } catch (err) {
    res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

export default router;

File: backend/app/api/v1/registry.py
"""Legacy registry endpoints for frontend backward-compatibility."""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.core.database import get_db_session
from app.models.vehicle import Vehicle
from app.models.medical import MedicalProfile, EmergencyContact

router = APIRouter(prefix="/registry")


@router.get("/vehicles")
async def get_vehicles(db: AsyncSession = Depends(get_db_session)):
    result = await db.execute(select(Vehicle))
    vehicles = result.scalars().all()
    return {"status": "SUCCESS", "data": [
        {
            "id": v.id, "name": v.name,
            "registrationNumber": v.registration_number,
            "deviceId": v.device_id,
            "driverName": v.driver_name,
            "bloodGroup": v.blood_group,
        }
        for v in vehicles
    ]}


@router.get("/medical")
async def get_medical(db: AsyncSession = Depends(get_db_session)):
    result = await db.execute(select(MedicalProfile).limit(1))
    profile = result.scalar_one_or_none()
    if not profile:
        return {"status": "SUCCESS", "data": None}
    return {"status": "SUCCESS", "data": {
        "fullName": profile.full_name,
        "age": profile.age,
        "gender": profile.gender,
        "bloodGroup": profile.blood_group,
        "abhaId": profile.abha_id,
        "allergies": profile.allergies,
        "medicalConditions": profile.medical_conditions,
        "primaryPhysicianName": profile.primary_physician_name,
        "primaryPhysicianPhone": profile.primary_physician_phone,
        "organDonor": profile.organ_donor,
    }}


@router.get("/contacts")
async def get_contacts(db: AsyncSession = Depends(get_db_session)):
    result = await db.execute(select(EmergencyContact))
    contacts = result.scalars().all()
    return {"status": "SUCCESS", "data": [
        {
            "id": c.id, "name": c.name, "relation": c.relation,
            "phone": c.phone, "isPrimary": c.is_primary, "notifySms": c.notify_sms,
        }
        for c in contacts
    ]}

